src/types/CodepointWidthDetector.cpp

```diff
@@ -325,68 +331,793 @@ namespace
+    0x02, 0x02, 0x43, 0x02, 0x40, 0x40, 0x40, 0x40,
+    0x80, 0x80, 0x80, 0x80, 0x02, 0x40, 0x40, 0x40,
+    0x80, 0x80, 0x80, 0x80, 0x40, 0x80, 0x80, 0x40,
+    0x40, 0x80, 0x40, 0x40, 0x40, 0x40, 0x40, 0x80,
+    0x80, 0x80, 0x40, 0x40, 0x80, 0x40, 0x40, 0x01,
+    0x01, 0x01, 0x01, 0x40, 0x40, 0x40, 0x40, 0x02,
+    0x02, 0x02, 0x01, 0x01, 0x01, 0x01, 0x01, 0x02,
+    0x02, 0x02, 0x02, 0x02, 0x40, 0x40, 0x02, 0x02,
+    0x02, 0x02, 0x02, 0x4c, 0x4c, 0x4c, 0x4c, 0x8c,
+    0x4c, 0x4c, 0x4c, 0xc0, 0xc0, 0xc0, 0x40, 0x40,
+    0x4c, 0x4c, 0x4c, 0xc0, 0xc0, 0xc0, 0xc0, 0xc0,
+    0xc0, 0x40, 0x4c, 0xc0, 0xc0, 0x40, 0x40, 0x4c,
+    0x4c, 0x4c, 0x4c, 0xcc, 0xcc, 0xc0, 0xc0, 0xc0,
+    0xc0, 0xc0, 0xc0, 0xcc, 0xcc, 0xc0, 0xc0, 0xc0,
+    0xc0, 0xc0, 0xc0, 0x8c, 0xc0, 0x8c, 0x8c, 0x8c,
+    0x8c, 0x8c, 0x8c, 0x8c, 0xc0, 0xc0, 0xc0, 0xc0,
+    0xc0, 0x4c, 0x4c, 0x4c, 0x4c, 0x4c, 0x4c, 0x45,
+    0x45, 0x45, 0x45, 0x45, 0x45, 0x45, 0x45, 0x80,
+    0x8c, 0x8c, 0x4c, 0x4c, 0x4c, 0x4c, 0x4c, 0x80,
+    0x80, 0x8c, 0x80, 0x80, 0x80, 0x80, 0x80, 0x8c,
+    0x8c, 0x8c, 0x8c, 0x8c, 0x8c, 0x80, 0x4c, 0x4c,
+    0x4c, 0x4c, 0x4c, 0x4c, 0x4c, 0x8c, 0x8c, 0x4c,
+    0x4c, 0x4c, 0x4c, 0x4c, 0x4c, 0x4c, 0x8c, 0x8c,
+    0x8c, 0x8c, 0x8c, 0x8c, 0x4c, 0x8c, 0x8c, 0x8c,
+    0x4c, 0x4c, 0x4c, 0x4c, 0x8c, 0x4c, 0x4c, 0x4c,
+    0x8c, 0x4c, 0x4c, 0x4c, 0x8c, 0x8c, 0x8c, 0x82,
+    0x82, 0x82, 0x82, 0x82, 0x8c, 0x4c, 0x8c, 0x8c,
+    0x8c, 0x8c, 0x8c, 0x8c, 0x4c, 0x4c, 0x8c, 0x8c,
+    0x8c, 0x8c, 0x8c, 0x8c, 0x40, 0x40, 0x4c, 0x4c,
+    0x4c, 0x8c, 0x8c, 0x8c, 0x8c, 0x4c, 0x40, 0x40,
+    0x40, 0x40, 0x4c, 0x4c, 0x4c, 0x4c, 0x4c, 0x4c,
+    0x40, 0x40, 0x40, 0x40, 0x8c, 0x8c, 0x8c, 0x8c,
+    0x40, 0x8c, 0x8c, 0x8c, 0x8c, 0x8c, 0x8c, 0x40,
+    0x8c, 0x41, 0x01, 0x41, 0x41, 0x41, 0x41, 0x41,
+    0x41, 0xc0, 0xc0, 0xc0, 0xc0, 0xc0, 0xc0, 0x40,
+    0x40,
+};
+static constexpr uint8_t s_joinRules[13][13] = {
+    0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0,
+    1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+    0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1,
+    0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0,
+    0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 0,
+    0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0,
+    0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0,
+    0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0,
+    0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0,
+    1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+    0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0,
+};
+[[msvc::forceinline]] constexpr uint8_t ucdLookup(const char32_t cp) noexcept
+{
+    const auto s1 = s_stage1[cp >> 11];
+    const auto s2 = s_stage2[s1 + ((cp >> 6) & 31)];
+    const auto s3 = s_stage3[s2 + ((cp >> 3) & 7)];
+    const auto s4 = s_stage4[s3 + ((cp >> 0) & 7)];
+    return s4;
+}
+[[msvc::forceinline]] constexpr uint8_t ucdGraphemeJoins(const uint8_t lead, const uint8_t trail) noexcept
 {
-    char32_t codepoint = 0;
+    const auto l = lead & 15;

```

Comment:
nit: personally i prefer masking constants to be hex, but if you prefer decimal that is OK too

Reply:
In this case I'm using `15` because that way I find it easier to see that it makes indexing into a `[16]` element large array safe.